web/html/src/manager/login/use-login-api.ts

```diff
@@ -0,0 +1,77 @@
+import { useEffect } from "react";
+import Network from "utils/network";
+import { useImmer } from "use-immer";
+
+type LoginApiStateType = {
+  messages: Array<string>;
+  success: boolean;
+  loading: boolean;
+};
+
+const errorsMessage = {
+  "error.invalid_login": t("Either the password or username is incorrect."),
+  "account.disabled": t("Your account has been deactivated."),
+  "error.user_readonly": t("This user has read only API access. WebUI login is denied."),
+};
+
+const useLoginApi = () => {
+  const [loginApiState, setLoginApiState] = useImmer<LoginApiStateType>({
+    messages: [],
+    success: false,
+    loading: false,
+  });
+
+  useEffect(
+    () =>
+      window.addEventListener("beforeunload", e => {
+        if (loginApiState.loading) {
+          const confirmationMessage = t("Are you sure you want to close this page while login is in progress?");
+          (e || window.event).returnValue = confirmationMessage;
+          return confirmationMessage;
+        }
+      }),
+    [loginApiState.loading]
```

Comment:
There seem to be some minor logic changes in this file. Have you tested those to make sure it doesn't break anything?

Reply:
I tested this but I think it will be better to come back to these later. We have a bunch of these across the codebase where we have empty hook dependency arrays, but we can pass those as a separate thing to avoid putting too many things in one pot.